substrate/frame/executive/src/lib.rs

```diff
@@ -551,53 +636,90 @@ where
 		sp_io::init_tracing();
 		sp_tracing::within_span! {
 			sp_tracing::info_span!("execute_block", ?block);
+			// Execute `on_runtime_upgrade` and `on_initialize`.
+			let mode = Self::initialize_block(block.header());
+			let num_inherents = Self::initial_checks(&block) as usize;
+			let (header, extrinsics) = block.deconstruct();
+			let num_extrinsics = extrinsics.len();
 
-			Self::initialize_block(block.header());
+			if mode == ExtrinsicInclusionMode::OnlyInherents && num_extrinsics > num_inherents {
+				// Invalid block
+				panic!("Only inherents are allowed in this block")
+			}
 
-			// any initial checks
-			Self::initial_checks(&block);
+			Self::apply_extrinsics(extrinsics.into_iter());
+			<frame_system::Pallet<System>>::note_finished_extrinsics();
 
-			// execute extrinsics
-			let (header, extrinsics) = block.deconstruct();
-			Self::execute_extrinsics_with_book_keeping(extrinsics, *header.number());
+			// In this case there were no transactions to trigger this state transition:
+			if !<frame_system::Pallet<System>>::inherents_applied() {
```

Comment:
Yea it would make more sense. We currently have:  
```rust
pub enum Phase {
	/// Applying an extrinsic.
	ApplyExtrinsic(u32),
	/// Finalizing the block.
	Finalization,
	/// Initializing the block.
	Initialization,
}
```
So i guess we could add an `ApplyInherent(u32)` as first phase after `Initialization`?

Reply:
Yes and `AfterInherent`